src/Lorentz/Contracts/BaseDAO/Proposal.hs

```diff
@@ -0,0 +1,682 @@
+-- SPDX-FileCopyrightText: 2020 TQ Tezos
+-- SPDX-License-Identifier: LicenseRef-MIT-TQ
+
+-- | Proposal related entrypoints
+module Lorentz.Contracts.BaseDAO.Proposal
+  ( propose
+  , proposalMetadata
+  , vote
+  , setVotingPeriod
+  , setQuorumThreshold
+  , flush
+  ) where
+
+import Lorentz
+import Lorentz.Contracts.BaseDAO.Management (authorizeAdmin)
+import Lorentz.Contracts.BaseDAO.FA2 (debitFrom, creditTo)
+import Lorentz.Contracts.BaseDAO.Types
+import Lorentz.Contracts.Spec.FA2Interface (TokenId)
+
+{-# ANN module ("HLint: ignore Reduce duplication" :: Text) #-}
+
+------------------------------------------------------------------------
+-- Helper
+------------------------------------------------------------------------
+
+checkIfProposalExist
+  :: forall store pm s. (NiceParameter pm, StorageC store pm)
+  => (ProposalKey : store : s) :-> (Proposal pm : s)
+checkIfProposalExist = do
+  stGet #sProposals
+  ifSome nop (failCustom_ #pROPOSAL_NOT_EXIST)
+
+ensureVotingPeriodIsNotOver
+  :: forall store pm s. (NiceParameter pm, StorageC store pm)
+  => (Proposal pm : store : s) :-> s
+ensureVotingPeriodIsNotOver = do
+  toField #pStartDate
+  swap
+  stToField #sVotingPeriod; int
+  -- push (60 * 60 * 24 :: Natural); -- one day
+  -- mul; -- multiply with the voting period
+  add; -- add to the start_date
+  toNamed #expectedEndDate
+  now; toNamed #currentDate
+  if #currentDate >=. #expectedEndDate then
+    failCustom_ #vOTING_PERIOD_OVER
+  else
+    nop
+
+ensureProposalIsUnique
+  :: forall store pm s. (NicePackedValue pm, StorageC store pm)
+  => ProposeParams pm : store : s :-> ProposeParams pm : store : s
+ensureProposalIsUnique = do
+  duupX @2; duupX @2
+  sender; swap; pair; pack @(ProposeParams pm, Address)
```

Comment:
1. I propose to define a type alias for data that is packed, otherwise it's easy to pass something wrong to `pack`. I mean, if you pass `(Address, ProposeParams)` instead of `(ProposeParams, Address)`, it will still compile. I would even go further and define something like `toProposalId` instruction and not use `pack` directly (outside of `toProposalId`).
2. I previously suggested that we also `blake2b` packed data, so that proposal IDs look more uniform and don't depend on proposal size. For large a proposal its ID can be too long.

Reply:
I've created a helper function `toProposalKey`  as well as add `blake2b` after `pack`.